spec/sec-properties-of-the-regexp-prototype-object-patch.html

```diff
@@ -122,121 +107,71 @@ <h1><ins>GetStringIndex ( _S_, _Input_, _e_ )</ins></h1>
-    <p>The value of the *name* property of this function is *"[Symbol.matchAll]"*.</p>
-
-    <emu-clause id="sec-createregexpstringiterator" aoid="CreateRegExpStringIterator">
-      <h1>CreateRegExpStringIterator ( _R_, _S_, _global_, _fullUnicode_<ins>, _matchOptions_</ins> )</h1>
-      <p>The abstract operation CreateRegExpStringIterator is used to create such iterator objects. It performs the following steps:</p>
+    <emu-clause id="sec-makegetindices" aoid="MakeGetIndices">
+      <h1><ins>MakeGetIndices ( _S_, _indices_ )</ins></h1>
       <emu-alg>
       1. Assert: Type(_S_) is String.
-      1. Assert: Type(_global_) is Boolean.
-      1. Assert: Type(_fullUnicode_) is Boolean.
-      1. <ins>Assert: _matchOptions_ is a MatchOptions Record.</ins>
-      1. Let _iterator_ be ObjectCreate(<emu-xref href="#table-properties-of-regexp-string-iterator-instances">%RegExpStringIteratorPrototype%</emu-xref>, &laquo; [[IteratingRegExp]], [[IteratedString]], [[Global]], [[Unicode]]<ins>, [[MatchOptions]]</ins>, [[Done]] &raquo;).
-      1. Set _iterator_.[[IteratingRegExp]] to _R_.
-      1. Set _iterator_.[[IteratedString]] to _S_.
-      1. Set _iterator_.[[Global]] to _global_.
-      1. Set _iterator_.[[Unicode]] to _fullUnicode_.
-      1. <ins>Set _iterator_.[[MatchOptions]] to _matchOptions_.</ins>
-      1. Set _iterator_.[[Done]] to *false*.
-      1. Return _iterator_.
+      1. Assert: _indices_ is a List where each element is either a Match Record or is *undefined*.
+      1. Let _steps_ be the steps of a GetIndices function as specified below.
+      1. Let _getter_ be CreateBuiltinFunction(_steps_, &laquo; [[InputString]], [[MatchIndicesList]], [[MatchIndicesArray]] &rqauo;).
+      1. Set _getter_.[[InputString]] to _S_.
+      1. Set _getter_.[[MatchIndicesList]] to _indices_.
+      1. Set _getter_.[[MatchIndicesArray]] to *undefined*.
+      1. Return _getter_.
+      </emu-alg>
+      <p>A <dfn>GetIndices</dfn> function is an anonymous built-in function with [[InputString]], [[MatchIndicesList]], and [[MatchIndicesArray]] internal slots. When a GetIndices function that expects no arguments is called it performs the following steps:</p>
+      <emu-alg>
+      1. Let _f_ be the active function object.
+      1. Let _A_ be _f_.[[MatchIndicesArray]].
+      1. If _A_ is not *undefined*, then
+        1. Return _A_.
+      1. Let _S_ be _f_.[[InputString]].
+      1. Let _indices_ be _f_.[[MatchIndicesList]].
+      1. Let _n_ be the number of elements in _indices_.
+      1. Assert: _n_ &lt; 2<sup>32</sup>-1.
+      1. Set _A_ to ! ArrayCreate(_n_).
+      1. Assert: The value of _A_'s `"length"` property is _n_.
+      1. If there exists an _element_ of _indices_ such that _element_ is a Match Record and _element_.[[GroupName]] is not *undefined*, then
+        1. Let _groups_ be ! ObjectCreate(*null*).
+      1. Else,
+        1. Let _groups_ be *undefined*.
+      1. Perform ! CreateDataProperty(_A_, `"groups"`, _groups_).
```

Comment:
it seems strange for something called GetIndices to be responsible for adding `groups`?

Reply:
It adds a `groups` to the indices array so that you can get the indices for a named capture group. In TS parlance:
```ts
// the result of RegExpBuiltinExec
interface RegExpExecArray extends Array<string> {
  index: number;
  input: string;
  groups?: { [key: string]: string; }
  readonly indices: RegExpExecIndicesArray;
}

// the array created by GetIndices
interface RegExpExecIndicesArray extends Array<[number, number]> {
  groups?: { [key: string]: [number, number]; };
}
```
In other words:

```ts
const m = re.exec(s);
m[0]; // a string
m.groups[name]; // a string
m.indices[0]; // a tuple of [start, end]
m.indices.groups[name]; // a tuple of [start, end]
```

Reply:
It has nothing to do with the `groups` on the array returned by RegExpBuiltinExec.

Reply:
Would a match either only be present in the indices array, or in the indices.groups object?

Reply:
The layout of the _substrings array_ (the normal result of `RegExpBuiltinExec`) and the layout of the _indices array_ (the `indices` property of the _substrings array_) are pretty much the same. Both the _substrings array_ and the _indices array_ have an element at index `0` for the entire match, elements at indices `1..n` for each capture group, and a `groups` property with named elements for each named capture group.

The only differences between these shared layouts is that the _substrings array_ has `string` elements, and the _indices array_ has `[number, number]` elements.